core/corehttp/gateway_handler_unixfs__redirects.go

```diff
@@ -0,0 +1,289 @@
+
+		// Fallback
+		webError(w, "ipfs resolve -r "+debugStr(contentPath.String()), err, http.StatusNotFound)
+		return nil, nil, false
+	}
+}
+
+func (i *gatewayHandler) handleRedirect(w http.ResponseWriter, r *http.Request, redirectRules []redirects.Rule) (bool, string, error) {
+	// Attempt to match a rule to the URL path, and perform the corresponding redirect or rewrite
+	pathParts := strings.Split(r.URL.Path, "/")
+	if len(pathParts) > 3 {
+		// All paths should start with /ipfs/cid/, so get the path after that
+		urlPath := "/" + strings.Join(pathParts[3:], "/")
+		rootPath := strings.Join(pathParts[:3], "/")
+		// Trim off the trailing /
+		urlPath = strings.TrimSuffix(urlPath, "/")
+
+		for _, rule := range redirectRules {
+			// get rule.From, trim trailing slash, ...
+			fromPath := urlpath.New(strings.TrimSuffix(rule.From, "/"))
+			match, ok := fromPath.Match(urlPath)
+			if !ok {
+				continue
+			}
+
+			// We have a match!  Perform substitutions.
+			toPath := rule.To
+			toPath = replacePlaceholders(toPath, match)
+			toPath = replaceSplat(toPath, match)
+
+			// Rewrite
+			if rule.Status == 200 {
+				// Prepend the rootPath
+				toPath = rootPath + toPath
+				return false, toPath, nil
+			}
+
+			// Or 404
+			if rule.Status == 404 {
+				toPath = rootPath + toPath
+				content404Path := ipath.New(toPath)
+				err := i.serve404(w, r, content404Path)
+				return true, toPath, err
+			}
+
+			// Or redirect
+			http.Redirect(w, r, toPath, rule.Status)
+			return true, toPath, nil
+		}
+	}
+
+	// No redirects matched
+	return false, "", nil
+}
+
+func replacePlaceholders(to string, match urlpath.Match) string {
+	if len(match.Params) > 0 {
+		for key, value := range match.Params {
+			to = strings.ReplaceAll(to, ":"+key, value)
+		}
+	}
+
+	return to
+}
+
+func replaceSplat(to string, match urlpath.Match) string {
+	return strings.ReplaceAll(to, ":splat", match.Trailing)
+}
+
+func (i *gatewayHandler) getRedirectRules(r *http.Request, redirectsFilePath ipath.Resolved) ([]redirects.Rule, error) {
+	// Convert the path into a file node
+	node, err := i.api.Unixfs().Get(r.Context(), redirectsFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("could not get _redirects node: %v", err)
+	}
+	defer node.Close()
+
+	// Convert the node into a file
+	f, ok := node.(files.File)
+	if !ok {
+		return nil, fmt.Errorf("could not convert _redirects node to file")
+	}
+
+	// Parse redirect rules from file
+	redirectRules, err := redirects.Parse(f)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse redirect rules: %v", err)
+	}
+
+	return redirectRules, nil
+}
+
+// Returns a resolved path to the _redirects file located in the root CID path of the requested path
+func (i *gatewayHandler) getRedirectsFile(r *http.Request, logger *zap.SugaredLogger) ipath.Resolved {
+	// r.URL.Path is the full ipfs path to the requested resource,
+	// regardless of whether path or subdomain resolution is used.
+	rootPath, err := getRootPath(r.URL.Path)
```

Comment:
nit: same as above, pass `contentPath` as an arg to `getRedirectsFile`, don't read `r.URL.Path` directly